### ISSUE TITLE
Is it normally?

### ISSUE THREAD

**Issue by avz-cmf** [mindplay-dk/jsonfreeze#6]
$serializer = new JsonSerializer();

$previous = new Exception('prop Exception1', 0);
$exc = new RuntimeException('prop RuntimeException', 0, $previous);

$valueExc = $serializer->unserialize($serializer->serialize($exc));
var_dump(null === $valueExc->getPrevious()); //true

$previous = new RuntimeException('prop RuntimeException', 0);
$exc = new Exception('prop Exception1', 0, $previous);

$valueExc = $serializer->unserialize($serializer->serialize($exc));
var_dump(null === $valueExc->getPrevious()); //false

**Comment by mindplay-dk** [comment]
I've never attempted to serialize exceptions.

I would not be suprised if, [like DateTime](https://www.mail-archive.com/internals@lists.php.net/msg86044.html), these types suffer from inconsistencies caused by implementation details in the backing C code.

If you can figure it out, you are welcome to submit a PR.

**Comment by avz-cmf** [comment]
Thnx for  the answer.

**Comment by avz-cmf** [comment]
But this problem is not specific for Ecxeption:

class A
{

    private $prop;

    public function set()
    {
        $this->prop = 1;
    }

    public function get()
    {
        return $this->prop;
    }

}

$a = new A();
$a->set();

class B extends A
{

}

$b = new B();
$b->set();

$serializer = new JsonSerializer();

$value = $serializer->unserialize($serializer->serialize($a));
var_dump(null === $value->get()); //false

$value = $serializer->unserialize($serializer->serialize($b));
var_dump(null === $value->get()); //true

**Comment by avz-cmf** [comment]
I think you can use next code for fix this problem:

```
$refClass = new \ReflectionClass('A');
$refProperty = $refClass->getProperty('prop');
$refProperty->setAccessible(true);
var_dump($refProperty->getValue($b)); //2
$refProperty->setValue($b, 3);
var_dump($refProperty->getValue($b)); //3
$refProperty->setAccessible(false);
```

**Comment by mindplay-dk** [comment]
Interesting, so `ReflectionClass::getProperties()` isn't consistent about inheritance - `public` and `protected` properties are available via immediate reflection, even if they were inherited, in fact they're reported as being properties of the class that inherited them; whereas `private` properties aren't reflected when inherited.

I've only ever used `protected` and `public` properties in models I've serialized, so I never ran into this.

The question is how to serialize inherited private properties?

Considering this case:

```php
class A { private $a = 1; }
class B extends A { private $a = 2; }
```

In this case, the fields `A::$a` and `B::$a` are different fields - even in instance of `B`, so we can't simply serialize inherited private properties using the property-name as key.

In the example above, we'd have to either serialize by qualifying *inherited* private properties with the class-name:

    {"#type": "B", "A#a": 1, "a": 2}

Or by qualifying *every* private property with the class-name:

    {"#type": "B", "A#a": 1, "B#a": 2}

What do you think?

**Comment by mindplay-dk** [comment]
You can try out the pre-release `0.3.3` branch now, which implements this:

    {"#type": "B", "A#a": 1, "a": 2}

Let me know how that works for you?

**Comment by avz-cmf** [comment]
0.3.3 is working.
But in real code {"#type": "B", "A#a": 1, "a": 2} will 
 {"#type": "B",
"a": 2
 "library-name\namespace1\namespace12\namespace3\ClassName#a": 1, 
} 
I think it is not good idea.
Anough:
 {"#type": "B",
0#"a": 2
1#a": 1, 
} 
Where 0# - point to #type": "B",  1# point to parent "B" level 1.
2# will parent of parent and  so on.

**Comment by avz-cmf** [comment]
P.S. May I open issue about problem with circular references?

**Comment by mindplay-dk** [comment]
Replacing class-names with numeric indices, in favor of saving bytes, is not something I want to do - keys like `"1#a"` are not human-readable, and transparency is one of the goals for this library.

> May I open issue about problem with circular references?

Circular references (per the overview in the README) are deliberately not supported, for reasons discussed [here](http://stackoverflow.com/questions/10489876/serialize-unserialize-php-object-graph-to-json) - you can find this link in the README as well. This is a deliberate choice, as serialization of graphics requires a very different data-structure and some serious trade-offs in terms of transparency.

This library is not a replacement for `serialize()` or `var_export()` - it's not intended for general-purpose serialization of any imaginable PHP object-graph, as such a library would need to make some serious trade-offs in favor of machine-friendly vs human-friendly data-structures.

In other words, general-purpose serialization to JSON, with support for references and circular graphs, would look like mish-mash to a person. JSON is a human-readable data-format, but that doesn't have much value if the data itself isn't intelligible to a person.

A better strategy is to design your models to avoid private fields and duplicate/circular references, or use private fields only for accidental state and disable serialization of private fields. That's the intent with this library.

**Comment by avz-cmf** [comment]
I see.
I have one more problem with this library.
I can define my own  serialization method for some class object, but I must know this class before call serialize. In my case I have serialize method for all classes of Exception. What I have to do? I have solved this, but code is not clean. You can see:
https://github.com/avz-cmf/zaboy/blob/master/src/utils/Json/Plugin/ExceptionSerializer.php
May be add callback (hook) for registration third party serializators?

**Comment by avz-cmf** [comment]
Also, note pls license type. I'm going to use (is it possibly) it in commercial project or in my own library.

**Comment by mindplay-dk** [comment]
> I can define my own serialization method for some class object, but I must know this class before call serialize.

Custom serialization must be defined ahead of time, e.g. using `defineSerialization()` - there is no scalable way to make that work for entire class hierarchies, which I expect is what you're looking for? This would require `instanceof` checks and more complex logic, which would slow down serialization of all objects.

It sounds like you're looking for a library that does general marshalling of PHP object graphs - like a JSON version of `serialize()` and `unserialize()`. That's not really what this library is. I use it as a component in libraries that provide JSON APIs etc. - but I use it as a component in the client/server implementations *internally* and apply it specifically to each case. It sounds like you're trying to use it more as a "framework" for something RPC-like?

That would require something more complex than this little library, something with quite a different set of qualities. You could definitely use this library as the basis for something like that, but it's beyond the scope of this library.

You may want to check out [fguillot/JsonRPC](https://github.com/fguillot/JsonRPC) which is something closer to an RPC client/server library - it does marshall Exceptions, and it follows the JSON-RPC 2 specification. It doesn't look like it does full object marshalling, so this probably isn't exactly the "framework" you're looking for either, but it might be closer.

> Also, note pls license type

Per the `composer.json` file, it's `LGPL-3.0+`

**Comment by avz-cmf** [comment]
Thank you.
